### PR TITLE
MM/EP-CHECK: added ep_check for memory mapper tls

### PR DESCRIPTION
**Pull request by hoopoepg**
- added implementation of ep_check for memory mapper
  based transports: posix & sysv
- updated tests

### CODE REVIEW

**Comment by brminich** [comment]
Error is relevant
```
[----------] 1 test from xpmem/test_uct_peer_failure_keepalive
[ RUN      ] xpmem/test_uct_peer_failure_keepalive.killed/0 <xpmem/memory>
[     INFO ] Testing component: xpmem
[swx-rdmz-ucx-legacy-01:28124:0:28124] Caught signal 7 (Bus error: nonexistent physical address)

/scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/sm/mm/base/mm_ep.c: [ uct_mm_ep_check() ]
      ...
      476     return uct_sm_ep_check(ep->fifo_ctl->owner.pid,
      477                            ep->fifo_ctl->owner.starttime, flags, comp);
      478 }

==== backtrace (tid:  28124) ====
 0 0x000000000005bd68 ucs_debug_print_backtrace()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/ucs/debug/debug.c:656
 1 0x00000000000197c2 uct_mm_ep_check()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../src/uct/sm/mm/base/mm_ep.c:479
 2 0x00000000006d8c68 test_uct_peer_failure_keepalive_killed_Test::test_body()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/uct/test_peer_failure.cc:595
 3 0x00000000005b4d06 ucs::test_base::run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:311
 4 0x00000000005b4d06 ucs::test_base::TestBodyProxy()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/test.cc:337
 5 0x0000000000596789 HandleSehExceptionsInMethodIfSupported<testing::Test, void>()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3562
 6 0x000000000058a9cd testing::Test::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3635
 7 0x000000000058aaa5 testing::TestInfo::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3812
 8 0x000000000058ac0f testing::TestCase::Run()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:3930
 9 0x000000000058f3bd testing::internal::UnitTestImpl::RunAllTests()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:5808
10 0x000000000058f6a0 testing::internal::UnitTestImpl::RunAllTests()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest-all.cc:5725
11 0x000000000052aa08 RUN_ALL_TESTS()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/gtest.h:20059
12 0x000000000052aa08 main()  /scrap/azure/agent-01/AZP_WORKSPACE/2/s/contrib/../test/gtest/common/main.cc:102
13 0x00000000000223d5 __libc_start_main()  ???:0
14 0x0000000000574d3f _start()  ???:0
=================================
[swx-rdmz-ucx-legacy-01:28124:0:28124] Process frozen...
make: *** [test] Terminated

```

**Comment by hoopoepg** [comment]
yep, xpmem transport was not covered by PR... now added

**Comment by hoopoepg** [comment]
io-demo failed to start

**Comment by hoopoepg** [comment]
@yosefe ok to squash?

**Comment by yosefe** [comment]
> @yosefe ok to squash?

yes

File: test/gtest/uct/test_peer_failure.cc [hunk 1]
```diff
@@ -557,7 +583,6 @@ UCS_TEST_SKIP_COND_P(test_uct_peer_failure_keepalive, killed,
     scoped_log_handler slh(wrap_errors_logger);
     flush();
     EXPECT_EQ(0, m_err_count);
-    flush();
```
Comment (brminich) [inline test/gtest/uct/test_peer_failure.cc:560]: unrelated?

File: test/gtest/uct/test_peer_failure.cc [hunk 2]
```diff
@@ -547,6 +555,24 @@ UCT_INSTANTIATE_TEST_CASE(test_uct_peer_failure_multiple)
 
 class test_uct_peer_failure_keepalive : public test_uct_peer_failure
 {
+public:
+    void kill_receiver()
+    {
+        /* hack: for SHM-based transports we can't really terminate
+         * peer EP, but instead we bit change process owner info to force
+         * ep_check failure. simulation of case when peer process is
```
Comment (brminich) [inline test/gtest/uct/test_peer_failure.cc:563]: minor: start with uppercase after `.`

File: test/gtest/uct/test_peer_failure.cc [hunk 3]
```diff
@@ -111,9 +113,15 @@ class test_uct_peer_failure : public uct_test {
         m_entities.push_back(m_receivers.back());
         m_sender->connect(m_receivers.size() - 1, *m_receivers.back(), 0);
 
-        am_handler_setter(this)(m_receivers.back());
-        /* Make sure that TL is up and has resources */
-        send_recv_am(m_receivers.size() - 1);
+        uct_iface_attr_t iface_attr;
+        ucs_status_t status = uct_iface_query(m_sender->ep(0)->iface,
```
Comment (brminich) [inline test/gtest/uct/test_peer_failure.cc:117]: can use `uct_test::entity::iface_attr()` instead

File: test/gtest/uct/test_peer_failure.cc [hunk 4]
```diff
@@ -111,9 +113,15 @@ class test_uct_peer_failure : public uct_test {
...
+                                              &iface_attr);
+        ASSERT_UCS_OK(status);
+        if (iface_attr.cap.flags & UCT_IFACE_FLAG_AM_SHORT) {
+            /* Make sure that TL is up and has resources */
+            am_handler_setter(this)(m_receivers.back());
+            send_recv_am(m_receivers.size() - 1);
+        }
```
Comment (brminich) [inline test/gtest/uct/test_peer_failure.cc:124]: is there a need to add check for others?

File: src/uct/sm/scopy/base/scopy_iface.h [hunk 5]
```diff
@@ -12,7 +12,7 @@
 #include <uct/sm/base/sm_iface.h>
 
 #define uct_scopy_trace_data(_tx) \
-    ucs_trace_data("%s [tx %p iov %zu/%zu length %zu/%zu] to %"PRIx64"(%+ld)", \
+    ucs_trace_data("%s [tx %p iov %zu/%zu length %zu/%zu] to %" PRIx64 "(%+ld)", \
```
Comment (brminich) [inline src/uct/sm/scopy/base/scopy_iface.h:15]: why needed?

File: src/uct/sm/scopy/base/scopy_ep.h [hunk 6]
```diff
@@ -44,9 +44,6 @@ typedef ucs_status_t
                           uint64_t remote_addr, uct_rkey_t rkey,
                           uct_scopy_tx_op_t tx_op);
 
-typedef struct uct_scopy_tx_arb_elem {
```
Comment (brminich) [inline src/uct/sm/scopy/base/scopy_ep.h:47]: unrelated?

File: src/ucs/sys/sys.c [hunk 7]
```diff
@@ -1439,3 +1439,21 @@ ucs_status_t ucs_sys_enum_threads(ucs_sys_enum_threads_cb_t cb, void *ctx)
...
 }
+
+ucs_status_t ucs_sys_get_process_starttime(pid_t pid, ucs_time_t *starttime)
+{
+    static const char *filename_fmt = "/proc/%d";
+    char filename[MAXPATHLEN];
+    struct stat stat_buf;
+    int res;
+
+    snprintf(filename, MAXPATHLEN, filename_fmt, (int)pid);
```
Comment (brminich) [inline src/ucs/sys/sys.c:1450]: can we create `filename` during endpoint creation (and cache it in the ep instead of pid)?

**Review by brminich — COMMENTED** [review]
what is with knem?

File: src/ucs/sys/sys.c (hunk 7 above)
Comment (hoopoepg) [reply src/ucs/sys/sys.c:1450]: technically yes, but it is more memory consuming, and there will not be way to get process starttime directly by pid - we have to get it as set of operations...

are we really want it?

File: src/uct/sm/scopy/base/scopy_ep.h (hunk 6 above)
Comment (hoopoepg) [reply src/uct/sm/scopy/base/scopy_ep.h:47]: related: this file is included into C++ gtest file & produces error

File: src/uct/sm/scopy/base/scopy_iface.h (hunk 5 above)
Comment (hoopoepg) [reply src/uct/sm/scopy/base/scopy_iface.h:15]: new C++ compiler require space between string literals

File: test/gtest/uct/test_peer_failure.cc (hunk 3 above)
Comment (hoopoepg) [reply test/gtest/uct/test_peer_failure.cc]: ok, done

File: test/gtest/uct/test_peer_failure.cc (hunk 2 above)
Comment (hoopoepg) [reply test/gtest/uct/test_peer_failure.cc]: fixed

File: test/gtest/uct/test_peer_failure.cc (hunk 4 above)
Comment (hoopoepg) [reply test/gtest/uct/test_peer_failure.cc:120]: may be, but later

File: test/gtest/uct/test_peer_failure.cc (hunk 1 above)
Comment (hoopoepg) [reply test/gtest/uct/test_peer_failure.cc:560]: not needed anymore

File: src/ucs/sys/sys.c (hunk 7 above)
Comment (brminich) [reply src/ucs/sys/sys.c:1450]: imo it would be better - we will have less intrusive check. We will have to increase ep size, not the address, should be ok for shm.
@yosefe, wdyt?

Comment (yosefe) [reply src/ucs/sys/sys.c:1450]: > imo it would be better - we will have less intrusive check. We will have to increase ep size, not the address, should be ok for shm.
> @yosefe, wdyt?

agree

File: src/ucs/sys/sys.c [hunk 8]
```diff
@@ -1439,3 +1439,26 @@ ucs_status_t ucs_sys_enum_threads(ucs_sys_enum_threads_cb_t cb, void *ctx)
...
+{
+    struct stat stat_buf;
+    int res;
+
+    res = stat(name, &stat_buf);
+    if (res != 0) {
+        return UCS_ERR_IO_ERROR; /* failed to get file info */
+    }
+
+    if (type == UCS_SYS_FILE_TIME_CTIME) {
```
Comment (yosefe) [inline src/ucs/sys/sys.c:1454]: maybe switch/case

File: src/ucs/sys/sys.h [hunk 9]
```diff
@@ -587,6 +596,18 @@ ucs_status_t ucs_sys_readdir(const char *path, ucs_sys_readdir_cb_t cb, void *ct
...
+ * Get file time
+ *
+ * @param [in]  name       File name
+ * @param [in]  type       Type of file time information
+ * @param [out] ctime      File creation time
+ *
+ * @return UCS_OK if file is found and got information.
+ */
+ucs_status_t ucs_sys_get_file_time(const char *name, ucs_sys_file_time_t type,
+                                   ucs_time_t *time);
```
Comment (yosefe) [inline src/ucs/sys/sys.h:610]: space line after

File: src/ucs/time/time.h [hunk 10]
```diff
@@ -75,6 +75,16 @@ static inline ucs_time_t ucs_time_from_sec(double sec)
...
 }
 
+
+/**
+ * Convert POSIX timespec to UCS time units.
+ */
+static inline ucs_time_t ucs_time_from_timespec(struct timespec *ts)
+{
+    return ucs_time_from_sec((double)ts->tv_sec +
+                             (double)ts->tv_nsec / UCS_NSEC_PER_SEC);
```
Comment (yosefe) [inline src/ucs/time/time.h:85]: can we do this without floating-point math?

File: src/uct/sm/base/sm_ep.c [hunk 11]
```diff
@@ -227,3 +228,28 @@ ucs_status_t uct_sm_ep_atomic_cswap32(uct_ep_h tl_ep, uint32_t compare,
...
+    UCT_CHECK_PARAM(flags == 0, "Unsupported flags: %u", flags);
+
+    status = ucs_sys_get_file_time(proc, UCS_SYS_FILE_TIME_CTIME, &createtime);
+    if (status != UCS_OK) { /* process is terminated */
+        return UCS_ERR_ENDPOINT_TIMEOUT;
+    }
+
+    /* if starttime is updated - then process is dead & pid reused */
+    return (starttime == createtime) ?
+           UCS_OK : UCS_ERR_ENDPOINT_TIMEOUT;
```
Comment (yosefe) [inline src/uct/sm/base/sm_ep.c:248]: if (status != OK || createtime!=startime) {
 return TIMEOUT;
}
return OK;

File: src/uct/sm/base/sm_ep.c [hunk 12]
```diff
@@ -227,3 +228,28 @@ ucs_status_t uct_sm_ep_atomic_cswap32(uct_ep_h tl_ep, uint32_t compare,
...
+}
+
+void uct_sm_ep_get_process_proc_dir(char *buffer, size_t max_len, pid_t pid)
+{
+    static const char *filename_fmt = "/proc/%d";
```
Comment (yosefe) [inline src/uct/sm/base/sm_ep.c:253]: no need for filename_fmt vat

File: src/uct/sm/base/sm_ep.h [hunk 13]
```diff
@@ -40,5 +43,8 @@ ucs_status_t uct_sm_ep_atomic32_fetch(uct_ep_h ep, uct_atomic_op_t opcode,
                                       uint32_t value, uint32_t *result,
                                       uint64_t remote_addr, uct_rkey_t rkey,
                                       uct_completion_t *comp);
+ucs_status_t uct_sm_ep_check(const char *proc, ucs_time_t starttime,
+                             unsigned flags, uct_completion_t *comp);
```
Comment (yosefe) [inline src/uct/sm/base/sm_ep.h:47]: space before/after

File: src/uct/sm/mm/base/mm_ep.h [hunk 14]
```diff
@@ -43,6 +44,13 @@ typedef struct uct_mm_ep {
         struct sockaddr_un     sockaddr;  /* address of signaling socket */
         socklen_t              addrlen;   /* address length of signaling socket */
     } signal;
+
+    /* owner of segment process information. we have to cache this value
+     * because some transports terminate segment when process gone (xpmem) */
+    struct {
+        char                  proc[UCT_SM_PROC_DIR_NAME_MAX_LEN]; /* Process owner proc dir */
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_ep.h:51]: maybe the proc path could be allocated dynamically, and created only if err handling is needed?
also this should be moved to sm/base

File: src/uct/sm/mm/base/mm_iface.h [hunk 15]
```diff
@@ -108,6 +108,10 @@ typedef struct uct_mm_fifo_ctl {
 
     /* 2nd cacheline */
     volatile uint64_t         tail;           /* How much was consumed */
+    struct {
+        pid_t                 pid;            /* Process owner pid */
+        ucs_time_t            starttime;      /* Process starttime */
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_iface.h:113]: why need starttime both on ep and iface?

File: src/uct/sm/scopy/cma/cma_ep.c [hunk 16]
```diff
@@ -93,3 +104,20 @@ ucs_status_t uct_cma_ep_tx(uct_ep_h tl_ep, const uct_iov_t *iov, size_t iov_cnt,
...
+
+ucs_status_t
+uct_cma_ep_check(uct_ep_h tl_ep, unsigned flags, uct_completion_t *comp)
+{
+    uct_cma_ep_t *ep   = ucs_derived_of(tl_ep, uct_cma_ep_t);
+    uct_iface_h  iface = ep->super.super.super.iface;
+    ucs_status_t status;
+
+    status = uct_sm_ep_check(ep->proc,
+                             ep->starttime, flags, comp);
```
Comment (yosefe) [inline src/uct/sm/scopy/cma/cma_ep.c:116]: wrap line

**Review by yosefe — COMMENTED** [review]
what about knem?

File: src/ucs/sys/sys.c (hunk 8 above)
Comment (hoopoepg) [reply src/ucs/sys/sys.c]: ok, re-implemented as switch/case

File: src/ucs/sys/sys.h (hunk 9 above)
Comment (hoopoepg) [reply src/ucs/sys/sys.h:610]: added space line

File: src/ucs/time/time.h (hunk 10 above)
Comment (hoopoepg) [reply src/ucs/time/time.h:85]: no: transformation is based on ```ucs_time_sec_value``` call which is floating-point

File: src/uct/sm/base/sm_ep.c (hunk 11 above)
Comment (hoopoepg) [reply src/uct/sm/base/sm_ep.c]: fixed

File: src/uct/sm/base/sm_ep.c (hunk 12 above)
Comment (hoopoepg) [reply src/uct/sm/base/sm_ep.c]: removed format var

File: src/uct/sm/base/sm_ep.h (hunk 13 above)
Comment (hoopoepg) [reply src/uct/sm/base/sm_ep.h:48]: added space lines

File: src/uct/sm/mm/base/mm_ep.h (hunk 14 above)
Comment (hoopoepg) [reply src/uct/sm/mm/base/mm_ep.h:51]: there is no info about err handling - err handling is UCP level.
for now not all shm-based transports have keepalive feature that is why starttime is not moved into sm/base object.
do you think it should be moved?

File: src/uct/sm/mm/base/mm_iface.h (hunk 15 above)
Comment (hoopoepg) [reply src/uct/sm/mm/base/mm_iface.h:113]: for sysv/posix/xpmem owner of segment stores own starttime in fifo control segment, later EP uses this info to check process alive. starttime is copied into EP because in case of xpmem when iface process is terminated it makes fifo memory unavailable and access to this memory produces seg.fault.

File: src/uct/sm/scopy/cma/cma_ep.c (hunk 16 above)
Comment (hoopoepg) [reply src/uct/sm/scopy/cma/cma_ep.c]: unwrapped

**Comment by hoopoepg** [comment]
> what about knem?

knem requires increase address size, if we ok with this - I will add it in separate PR

File: src/uct/sm/mm/base/mm_ep.c [hunk 17]
```diff
@@ -180,6 +178,10 @@ static UCS_CLASS_CLEANUP_FUNC(uct_mm_ep_t)
     uct_mm_iface_t  *iface = ucs_derived_of(self->super.super.iface, uct_mm_iface_t);
     uct_mm_remote_seg_t remote_seg;
 
+    if (self->keepalive != NULL) {
+        ucs_free(self->keepalive);
+    }
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_ep.c:183]: ucs_free can handle NULL, no need to check

File: src/uct/sm/mm/base/mm_ep.c [hunk 18]
```diff
@@ -479,10 +481,25 @@ uct_mm_ep_check(uct_ep_h tl_ep, unsigned flags, uct_completion_t *comp)
...
+
+    status = uct_sm_ep_check(ep->keepalive->proc, ep->keepalive->starttime,
+                             flags, comp);
     if (status != UCS_OK) {
-        uct_set_ep_failed(&UCS_CLASS_NAME(uct_mm_ep_t), &ep->super.super,
-                          iface, status);
+failed:
+        return uct_set_ep_failed(&UCS_CLASS_NAME(uct_mm_ep_t), &ep->super.super,
+                                 iface, status);
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_ep.c:502]: doesn't look very nice to jump inside a block
let's do like this:
```
    status = uct_sm_ep_check(....)
    if (status != UCS_OK) {
        goto err_set_ep_failed;
    }

    return UCS_OK;

err_set_ep_failed:
    uct_set_ep_failed(...)
    return status;
```

File: src/uct/sm/mm/base/mm_ep.h [hunk 19]
```diff
@@ -18,6 +18,13 @@ KHASH_INIT(uct_mm_remote_seg, uintptr_t, uct_mm_remote_seg_t, 1,
            kh_int64_hash_func, kh_int64_hash_equal)
 
 
+/* owner of segment process information. we have to cache this value
+ * because some transports terminate segment when process gone (xpmem) */
+typedef struct uct_mm_keepalive_info {
+    char       proc[UCT_SM_PROC_DIR_NAME_MAX_LEN]; /* Process owner proc dir */
+    ucs_time_t starttime;                          /* Process starttime */
```
Comment (yosefe) [inline src/uct/sm/mm/base/mm_ep.h:25]: could do like this:
typedef struct uct_mm_keepalive_info {
 ucs_time_t start_time;
 char prof_file_path[0]
}
and then malloc according to actual size - and remove UCT_SM_PROC_DIR_NAME_MAX_LEN constant

Comment (hoopoepg) [reply src/uct/sm/mm/base/mm_ep.h]: ok, done

**Review by yosefe — APPROVED** [review]
pls squash
